src/sparseml/pytorch/optim/manager.py

```diff
@@ -32,7 +31,185 @@
+        self._wrapped_epoch = epoch
+        self._wrapped_steps = round(epoch * steps_per_epoch)
+
+    def __del__(self):
+        try:
+            # propagate delete to manager so it can disconnect hooks and objects
+            del self._wrapped_manager
+        except Exception:
+            pass
+
+    def __getattr__(self, item):
+        if item in self.__dict__:
+            return getattr(self, item)
+
+        return getattr(self._wrapped, item)
+
+    def __setattr__(self, key, value):
+        if key.startswith("_wrapped"):
+            super().__setattr__(key, value)
+        else:
+            setattr(self._wrapped, key, value)
+
+    @property
+    def wrapped(self):
+        """
+        :return: The object to wrap the step function and properties for.
+        """
+        return self._wrapped
+
+    @property
+    def wrapped_manager(self):
+        """
+        :return: The manager to forward lifecycle calls into such as step.
+        """
+        return self._wrapped_manager
+
+    @property
+    def wrapped_optimizer(self) -> Any:
+        """
+        :return: The optimizer used in the training process.
+        """
+        return self._wrapped_optimizer
+
+    @property
+    def wrapped_module(self) -> Module:
+        """
+        :return: The model/module used in the training process.
+        """
+        return self._wrapped_module
+
+    @property
+    def wrapped_steps_per_epoch(self) -> int:
+        """
+        :return: The number of optimizer steps (batches) in each epoch.
+        """
+        return self._wrapped_steps_per_epoch
+
+    @property
+    def wrapped_epoch(self) -> float:
+        """
+        :return: The current epoch the wrapped object is at.
+        """
+        return self._wrapped_epoch
+
+    @property
+    def wrapped_steps(self) -> int:
+        """
+        :return: The current number of steps that have been called for
+            the wrapped object.
+        """
+        return self._wrapped_steps
+
+    def step(self, *args, **kwargs):
+        """
+        Override for the step function.
+        Calls into the base step function with the args and kwargs.
+
+        :param args: Any args to pass to the wrapped objects step function.
+        :param kwargs: Any kwargs to pass to the wrapped objects step function.
+        :return: The return, if any, from the wrapped objects step function
+        """
+        if self._wrapped_manager.enabled:
+            self._wrapped_manager.update(
+                self._wrapped_module,
+                self._wrapped_optimizer,
+                self._wrapped_epoch,
+                self._wrapped_steps_per_epoch,
+            )
+            self._wrapped_manager.optimizer_pre_step(
+                self._wrapped_module,
+                self._wrapped_optimizer,
+                self._wrapped_epoch,
+                self._wrapped_steps_per_epoch,
+            )
+
+        ret = self._wrapped.step(*args, **kwargs)
```

Comment:
Can you clarify what the difference between the self._wrapped and the self._wrapped_optimizer? In the context of amp support, what would we pass in for the _wrapped?

Reply:
Wrapped is what we override the step call for. Wrapped optimizer is the pytorch optimizer implementation needed to be passed through to the modifiers

Reply:
I saw we passed the same optimizer into both the _wrapped and _wrapped_optimizer (in optimizer.py). This is not always the case, right? (otherwise, we could just use _wrapped_optimizer)